mlir/lib/Gradient/Transforms/ConversionPatterns.cpp

```diff
@@ -375,6 +493,162 @@ struct BackpropOpPattern : public ConvertOpToLLVMPattern<BackpropOp> {
         return bufferSizeBytes;
     }
 
+    void convertCustomGradArgumentTypes(TypeRange argTypes,
+                                        SmallVectorImpl<Type> &llvmArgTypes) const
+    {
+        for (Type argType : argTypes) {
+            llvmArgTypes.append(2, argType);
+        }
+    }
+
+    func::FuncOp genAugmentedForward(func::FuncOp qnode, OpBuilder &builder) const
+    {
+        std::string augmentedName = (qnode.getName() + ".augfwd").str();
+        auto augmentedForward = SymbolTable::lookupNearestSymbolFrom<func::FuncOp>(
+            qnode, builder.getStringAttr(augmentedName));
+        if (augmentedForward) {
+            return augmentedForward;
+        }
+        assert(qnode.getNumResults() == 0 && "Expected QNode to be in destination-passing style");
+        MLIRContext *ctx = builder.getContext();
+        OpBuilder::InsertionGuard insertionGuard(builder);
+        builder.setInsertionPointAfter(qnode);
+        // The tape type is a null pointer because we don't need to pass any data from the forward
+        // pass to the reverse pass.
+        auto tapeType = LLVM::LLVMPointerType::get(ctx);
+        SmallVector<Type> argTypes;
+        convertCustomGradArgumentTypes(qnode.getArgumentTypes(), argTypes);
+        augmentedForward = builder.create<func::FuncOp>(
+            qnode.getLoc(), augmentedName, FunctionType::get(ctx, argTypes, {tapeType}));
+        augmentedForward.setPrivate();
+        Location loc = qnode.getLoc();
+
+        Block *entry = augmentedForward.addEntryBlock();
+        builder.setInsertionPointToStart(entry);
+
+        // Every other argument is a shadow
+        SmallVector<Value> arguments;
+        for (unsigned i = 0; i < qnode.getNumArguments(); i++) {
+            arguments.push_back(augmentedForward.getArgument(i * 2));
+        }
+
+        builder.create<func::CallOp>(loc, qnode, arguments);
+        Value tape = builder.create<LLVM::NullOp>(loc, tapeType);
+        builder.create<func::ReturnOp>(loc, tape);
+        return augmentedForward;
+    }
+
+    // `qnodeType` is the original type of the function prior to conversion to destination-passing
+    // style and wrapping memrefs into pointers.
+    func::FuncOp genCustomQGradient(func::FuncOp qnode, Location loc, func::FuncOp qgradFn,
```

Comment:
A small doc would be nice for this function as it is the backbone of the new gradient lowering